libs/vkd3d/command.c

```diff
@@ -1259,6 +1278,290 @@ HRESULT d3d12_fence_create(struct d3d12_device *device,
+        d3d12_shared_fence_dec_ref(fence);
+        d3d12_device_release(device);
+    }
+
+    return refcount;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_GetPrivateData(d3d12_fence_iface *iface,
+        REFGUID guid, UINT *data_size, void *data)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+
+    TRACE("iface %p, guid %s, data_size %p, data %p.\n",
+            iface, debugstr_guid(guid), data_size, data);
+
+    return vkd3d_get_private_data(&fence->private_store, guid, data_size, data);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_SetPrivateData(d3d12_fence_iface *iface,
+        REFGUID guid, UINT data_size, const void *data)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+
+    TRACE("iface %p, guid %s, data_size %u, data %p.\n",
+            iface, debugstr_guid(guid), data_size, data);
+
+    return vkd3d_set_private_data(&fence->private_store, guid, data_size, data,
+            NULL, NULL);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_SetPrivateDataInterface(d3d12_fence_iface *iface,
+        REFGUID guid, const IUnknown *data)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+
+    TRACE("iface %p, guid %s, data %p.\n", iface, debugstr_guid(guid), data);
+
+    return vkd3d_set_private_data_interface(&fence->private_store, guid, data,
+            NULL, NULL);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_GetDevice(d3d12_fence_iface *iface, REFIID iid, void **device)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+
+    TRACE("iface %p, iid %s, device %p.\n", iface, debugstr_guid(iid), device);
+
+    return d3d12_device_query_interface(fence->device, iid, device);
+}
+
+static UINT64 STDMETHODCALLTYPE d3d12_shared_fence_GetCompletedValue(d3d12_fence_iface *iface)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+    const struct vkd3d_vk_device_procs *vk_procs = &fence->device->vk_procs;
+    uint64_t completed_value;
+    VkResult vr;
+
+    TRACE("iface %p\n", iface);
+
+    vr = VK_CALL(vkGetSemaphoreCounterValueKHR(fence->device->vk_device, fence->timeline_semaphore, &completed_value));
+    if (vr != VK_SUCCESS)
+    {
+        ERR("Failed to get shared fence counter value, error %d.\n", vr);
+        return 0;
+    }
+    return completed_value;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_SetEventOnCompletion(d3d12_fence_iface *iface,
+        UINT64 value, HANDLE event)
+{
+    struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
+    const struct vkd3d_vk_device_procs *vk_procs = &fence->device->vk_procs;
+    VkSemaphoreWaitInfo wait_info;
+    VkResult vr;
+
+    TRACE("iface %p, value %#"PRIx64", event %p.\n", iface, value, event);
+
+    if (event)
+    {
+        FIXME("Signaling events on shared fence completion not supported.\n");
```

Comment:
I would guess this is theoretically possible to support if we make a WINE vendor extension or Win32-only EXT? Ala `vkSemaphoreSetWin32EventOnCompletionEXT`?

Reply:
Yeah, if we run into this either of those two options should work.  DXVK uses a wait thread for ID3D11Fence, but the problem is that if you only want one thread per fence, it's not easy to support the rewind case.  Really the only working option would be a threadpool where you have a new worker for every wait, or as you mentioned a wine extension which would be pretty trivial.

Reply:
I'm not too concerned as long as we have a path forward if need be.